wm_dll.c

```diff
@@ -13,20 +18,61 @@ __declspec(dllexport) LRESULT CALLBACK ShellProc(int code, WPARAM wparam, LPARAM
 	return CallNextHookEx(NULL, code, wparam, lparam);
 }
 
+__declspec(dllexport) LRESULT HotkeyProc(int code, WPARAM wparam, LPARAM lparam) {
+	DEBUG_PRINT("HotkeyProc called - %i %i %i", code, wparam, lparam); 
+	
+	switch(wparam) 
+	{ 
+		//TODO Can either trigger an event like the ShellProc callback, or handle directly. 
+		// one method to handle virtual desktops is using the IVirtualDesktopManager in ShObjIdl but 
+		// that is only available for Window 10 1809 or later. 
+		case WORKSPACE_1:
+			puts("Switch to workspace 1");
+			break;
+		case WORKSPACE_2:
+			puts("Switch to workspace 2"); 
+			break;
+		case WORKSPACE_3:
+			puts("Switch to workspace 3"); 
+			break;
+		case WORKSPACE_4:
+			puts("Switch to workspace 4"); 
+			break;
+		case WINDOW_UP: 
+			puts("Highlight window above"); 
+			break; 
+		case WINDOW_DOWN: 
+			puts("Highlight window below"); 
+			break; 
+		case WINDOW_LEFT: 
+			puts("Highlight window left"); 
+			break; 
+		case WINDOW_RIGHT: 
+			puts("Highlight window right"); 
+			break; 
+		default: 
+			DEBUG_PRINT("Unhandled hotkey message! Hotkey ID: %i", wparam); 
+			break; 
+	}
+	
+	return CallNextHookEx(NULL, code, wparam, lparam);
```

Comment:
Why are you calling CallNextHook? this function is not a hook, it is only a handler for the hot key message

Reply:
Accidentally left that in since that function was originally a hook.